gson/src/test/java/com/google/gson/regression/JsonAdapterNullSafeTest.java

```diff
@@ -0,0 +1,186 @@
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonToken;
+import com.google.gson.stream.JsonWriter;
+
+import junit.framework.TestCase;
+
+public class JsonAdapterNullSafeTest extends TestCase {
+  private final Gson gson = new Gson();
+
+  /**
+   * The recursiveCall in {@link ControlData.JsonAdapterFactory} is the source of this bug
+   * because it returns a null type adapter.
+   */
+  public void testTypeAdapterFactoryNullSafeBug() throws IOException {
+      ControlData control = new ControlData("ec57803e", 2, true, 11211);
+      Device device = new Device("device1", control);
+      String json = gson.toJson(device);
+
+      json = "\"{\\\"id\\\":\\\"ec57803e2\\\",\\\"category\\\":2,\\\"alwaysOn\\\":true,\\\"codeset_id\\\":11211}\"";
+      control = gson.fromJson(json, ControlData.class);
+      assertEquals("ec57803e2", control.id);
+      assertTrue(control.alwaysOn);
+      assertEquals(11211, control.codesetId);
+      assertEquals(2, control.category);
+
+      String deviceJson = "{'id':'device1','controlData':null}";
+      device = gson.fromJson(deviceJson, Device.class);
+      assertNull(device.controlData);
+
+      deviceJson = "{'id':'device1','controlData':{'id':'ec57803e2','category':2,'alwaysOn':true,'codeset_id':12221}}";
+      device = gson.fromJson(deviceJson, Device.class);
+      assertEquals(12221, device.controlData.codesetId);
+
+      deviceJson = "{'id':'device1','controlData':'\\\"{}\\\"'}";
+      device = gson.fromJson(deviceJson, Device.class);
+
+      try {
+          deviceJson = "{'id':'device1','controlData':'a'}";
+          device = gson.fromJson(deviceJson, Device.class);
+          assertNotNull(device.controlData);
+          fail();
+      } catch (JsonSyntaxException expected) {}
+
+      deviceJson = "{'id':'device1','controlData':'  '}";
+      device = gson.fromJson(deviceJson, Device.class);
+      assertNull(device.controlData);
+  }
+
+  private static final class Device {
+    @SuppressWarnings("unused")
+    String id;
+    ControlData controlData;
+
+    public Device(String id, ControlData controlData) {
+      this.id = id;
+      this.controlData = controlData;
+    }
+  }
+
+  @JsonAdapter(ControlData.JsonAdapterFactory.class)
+  private static final class ControlData {
+    String id;
+    int category;
+    boolean alwaysOn;
+    @SerializedName("codeset_id") int codesetId;
+    ControlData(String id, int category, boolean alwaysOn, int codesetId) {
+      this.id = id;
+      this.category = category;
+      this.alwaysOn = alwaysOn;
+      this.codesetId = codesetId;
+    }
+
+    /**
+     * DeviceControlData is received as String in JSON instead of proper JSON.
+     * So, we need to write a special type adapter.
+     */
+    static final class JsonAdapterFactory extends StringifiedJsonAdapterFactory<ControlData> {
+      private static final ThreadLocal<Boolean> recursiveCall = new ThreadLocal<Boolean>();
```

Comment:
Yes. As I debugged the problem, I realized that the constructorConstructor.get() resulted in a recursive call to the JsonAdapterFactory.  Since we don't want to be in an infinite loop, I had used a ThreadLocal to avoid that problem in Gson 2.5.
Now, that ThreadLocal returning null was causing this regression.